src/replication.c

```diff
@@ -1914,6 +1943,169 @@ disconnect_nonresponding_replica(replica_t *replica, uint64_t io_seq,
+		rcomm_mgmt->cmds_failed++;
+		free(snap_details);
+		failed = true;
+	} else {
+		rcomm_mgmt->cmds_succeeded++;
+	}
+
+	if ((rcomm_mgmt->caller_gone == 1) &&
+	    (rcomm_mgmt->cmds_sent == (rcomm_mgmt->cmds_failed + rcomm_mgmt->cmds_succeeded)))
+		delete = true;
+
+	if (failed)
+		goto out;
+
+	for (i = 0, assigned = false;
+	    i < (rcomm_mgmt->buf_size / sizeof (*snap_resp_array));
+		i++) {
+		if (snap_resp_array[i] == NULL) {
+			snap_resp_array[i] = (struct zvol_snapshot_list *)snap_details;
+			assigned = true;
+			break;
+		}
+	}
+
+	ASSERT(assigned);
+
+out:
+
+	MTX_UNLOCK(&rcomm_mgmt->mtx);
+
+	if (delete == true) {
+		DESTROY_SNAPSHOT_RESP_LIST(rcomm_mgmt);
+		free_rcommon_mgmt_cmd(rcomm_mgmt);
+	}
+}
+
+static char *
+get_replica_id_for_guid(spec_t *spec, uint64_t guid)
+{
+	replica_t *replica;
+	char *replica_id = NULL;
+
+	TAILQ_FOREACH(replica, &spec->rq, r_next) {
+		if (replica->zvol_guid == guid) {
+			replica_id = xstrdup(replica->replica_id);
+			break;
+		}
+	}
+
+	return replica_id;
+}
+
+static void
+process_snaplist_response(spec_t *spec, char **resp, void *buf, int rcount)
+{
+	struct zvol_snapshot_list **snap_resp_array = (struct zvol_snapshot_list **)buf;
+	int i = 0;
+	uint64_t total_len;
+	char *msg = NULL, *replica_id = NULL;
+	const char *json_string = NULL;
+	struct json_object *jobj, *jarray, *resp_obj, *snap_obj;
+
+	jarray = json_object_new_array();
+
+	for (i = 0; i < rcount; i += 1) {
+		if (snap_resp_array[i]) {
+			resp_obj = snap_obj = NULL;
+			jobj = json_object_new_object();
+
+			replica_id = get_replica_id_for_guid(spec, snap_resp_array[i]->zvol_guid);
+			if(!replica_id) {
+				REPLICA_ERRLOG("No replica_id for replica:%lu", snap_resp_array[i]->zvol_guid);
+				continue;
+			}
+
+			json_object_object_add(jobj, "replica_id", json_object_new_string(replica_id));
+			if (snap_resp_array[i]->error)
+				REPLICA_ERRLOG("SNAP_LIST recieved an error(%d) for replica(%s)\n", snap_resp_array[i]->error, replica_id);
+
+			free(replica_id);
+
+			resp_obj = json_tokener_parse(snap_resp_array[i]->data);
+			(void) json_object_object_get_ex(resp_obj,
+			    "snapshot", &snap_obj);
+			json_object_object_add(jobj, "snapshot", snap_obj);
+
+			json_object_array_add(jarray, jobj);
+		}
+	}
+	jobj = json_object_new_object();
+	json_object_object_add(jobj, "snapshots", jarray);
+
+	json_string = json_object_to_json_string_ext(jobj, JSON_C_TO_STRING_PLAIN);
+	total_len = strlen(json_string) + 1;
+	msg = malloc(total_len);
+	memset(msg, 0, total_len);
+	strncpy(msg, json_string, total_len - 1);
+	json_object_put(jobj);
+	*resp = msg;
+}
+
+char *
+istgt_lu_fetch_snaplist(spec_t *spec, int wait_time, char *snapname)
+{
+	replica_t *replica;
+	int free_rcomm_mgmt = 0, success;
+	struct timespec last;
+	rcommon_mgmt_cmd_t *rcomm_mgmt;
+	char *response = NULL;
+	int num_replica = 0;
+
+	clock_gettime(CLOCK_MONOTONIC, &last);
+
+	REPLICA_LOG("fetching snapshot list for %s@%s", spec->volname, snapname);
```

Comment:
lets have a similar log at the end of function as well

Reply:
done